src/riot-rs-threads/src/sync/mutex.rs

```diff
@@ -0,0 +1,193 @@
+use core::{
+    cell::UnsafeCell,
+    ops::{Deref, DerefMut},
+};
+
+use critical_section::CriticalSection;
+use riot_rs_runqueue::{RunqueueId, ThreadId};
+
+use crate::{thread::ThreadState, threadlist::ThreadList, THREADS};
+
+/// A basic mutex with priority inheritance.
+pub struct Mutex<T> {
+    state: UnsafeCell<LockState>,
+    inner: UnsafeCell<T>,
+}
+
+unsafe impl<T> Sync for Mutex<T> {}
+
+/// State of a [`Mutex`].
+enum LockState {
+    Unlocked,
+    Locked {
+        // The current owner of the lock.
+        owner_id: ThreadId,
+        // The original priority of the current owner (without priority inheritance).
+        owner_prio: RunqueueId,
+        // Waiters for the mutex.
+        waiters: ThreadList,
+    },
+}
+
+impl LockState {
+    /// Returns a [`LockState::Locked`] with the current thread as the owner
+    /// and an empty waitlist.
+    ///
+    /// # Panics
+    ///
+    /// Panics if called outside of a thread context.
+    fn locked_with_current(cs: CriticalSection) -> Self {
+        let (owner_id, owner_prio) = THREADS.with_mut_cs(cs, |mut threads| {
+            let current = threads
+                .current()
+                .expect("Function should be called inside a thread context.");
+            (current.pid, current.prio)
+        });
+        LockState::Locked {
+            waiters: ThreadList::new(),
+            owner_id,
+            owner_prio,
+        }
+    }
+}
+
+impl<T> Mutex<T> {
+    /// Creates a new **unlocked** [`Mutex`].
+    pub const fn new(value: T) -> Self {
+        Self {
+            state: UnsafeCell::new(LockState::Unlocked),
+            inner: UnsafeCell::new(value),
+        }
+    }
+
+    /// Returns whether the mutex is locked.
+    pub fn is_locked(&self) -> bool {
+        critical_section::with(|_| {
+            let state = unsafe { &*self.state.get() };
+            !matches!(state, LockState::Unlocked)
+        })
+    }
+
+    /// Acquires a mutex, blocking the current thread until it is able to do so.
+    ///
+    /// If the mutex was unlocked, it will be locked and a [`MutexGuard`] is returned.
+    /// If the mutex is locked, this function will block the current thread until the mutex gets
+    /// unlocked elsewhere.
+    ///
+    /// If the current owner of the mutex has a lower priority than the current thread, it will inherit
+    /// the waiting thread's priority.
+    /// The priority is reset once the mutex is released. This means that a **user can not change a thread's
+    /// priority while it holds the lock**, because it will be changed back after release!
+    ///
+    /// # Panics
+    ///
+    /// Panics if called outside of a thread context.
+    pub fn lock(&self) -> MutexGuard<T> {
+        critical_section::with(|cs| {
+            let state = unsafe { &mut *self.state.get() };
```

Comment:
> I'm not yet convinced that the safety conditions for casting the pointer obtained from that UnsafeCell to a &mut are verified here (as that lock() method only takes a shared reference).

I am not sure if I understand you correctly. Do you mean safe in the sense that no other threads have mutual access at the same time? This is ensured because all mutual access only every happens inside global critical sections.